### ISSUE TITLE
PIDSet thread safety

### ISSUE THREAD

**Issue by NathanBaulch** [AsynkronIT/protoactor-go#229]
I'm experiencing fatal "concurrent map iteration and map write" errors due to race conditions in the mutable `PIDSet` type. This is very easy to reproduce:
```go
var s actor.PIDSet
go func() {
	i := 0
	for {
		i++
		s.Add(&actor.PID{Id: strconv.Itoa(i)})
		time.Sleep(time.Millisecond)
	}
}()
go func() {
	for {
		s.ForEach(func(_ int, _ actor.PID) {})
		time.Sleep(time.Millisecond)
	}
}()
select {}
```

In my particular case I have a broadcast group (`router.NewBroadcastGroup`) that receives a large number of `router.AddRoutee` and other messages simultaneously from different goroutines. `broadcastRouterState` exposes its `PIDSet` via the `GetRoutees` function, which in turn is used by `groupRouterActor` to both mutate and iterate over the PIDs.

Looking at other internal uses of `PIDSet`, I see types like `router.process` synchronize access to this type (see [here](https://github.com/AsynkronIT/protoactor-go/blob/b4e5ee3e19305b162ee4cd5e606cbb03a3cac48f/router/process.go#L52-L58)) whereas other types like `actor.localContext` and `remote.endpointWatcher` do not.

It seems like this could be fixed in a few different ways:
1. Switch the map field in `PIDSet` to a `sync.Map`
2. Synchronize `PIDSet` functions with `RWMutex` or similar
3. Make sure all `router.Interface` implementations return a clone of their internal PIDs from `GetRoutees`
4. Mirror the `Add`, `Remove` and `ForEach` functions in `router.Interface` and synchronize them there

Based on the existence and use of the `Interface.SetRoutees` function, I would lean towards the third option. So unless I've misunderstood something important, I'd be happy to submit a pull request.

**Comment by NathanBaulch** [comment]
FWIW, here's another similar program that reproduces the error specifically with a broadcast group:
```
grp := actor.Spawn(router.NewBroadcastGroup())
go func() {
	i := 0
	for {
		i++
		grp.Tell(&router.AddRoutee{&actor.PID{Id: strconv.Itoa(i)}})
		time.Sleep(10 * time.Millisecond)
	}
}()
go func() {
	for {
		grp.Tell(struct{}{})
		time.Sleep(10 * time.Millisecond)
	}
}()
select {}
```

**Comment by bryanpkc** [comment]
I agree with solution 3. `PIDSet` is almost always exclusively owned and used in a single-threaded context, and it doesn't make sense to add intrinsic synchronization to it.

**Comment by PotterDai** [comment]
@NathanBaulch It seems in go 1.8 there is enhanced concurrency check implemented as is mentioned here
https://stackoverflow.com/questions/42453442/concurrent-map-iteration-and-map-write-error-in-golang-1-8
@rogeralsing The problem is here:
https://github.com/AsynkronIT/protoactor-go/blob/dev/router/process.go#L23
Shall we just let route message go through mailbox as well, so that we can completely avoid race condition?

**Comment by rogeralsing** [comment]
>PIDSet is almost always exclusively owned and used in a single-threaded context,

agree, it was designed as the child container for actors.

**Comment by bryanpkc** [comment]
@PotterDai The fact that multiple non-management messages can be routed concurrently is key to the performance of routers. I don't think it makes sense to serialize messages intended for routees.

**Comment by PotterDai** [comment]
@bryanpkc but duplicate PIDSet for every message will affect performance as well?

**Comment by bryanpkc** [comment]
If I understand the design correctly, management messages like `AddRoutee`, `RemoveRoutee` and `BroadcastMessage` are intended to be rare, relative to the amount of messages being routed and the number of routees in use. You wouldn't need to duplicate the `PIDSet` for every user message.

**Comment by rogeralsing** [comment]
>broadcastRouterState exposes its PIDSet via the GetRoutees function, which in turn is used by groupRouterActor to both mutate and iterate over the PIDs.

What am I missing here, if the set is protected by the router management actor, 
Is there anything else than the actor mutating the set also?

**Comment by bryanpkc** [comment]
@rogeralsing The router management actor can modify the `PIDSet` while the routing logic is using it to route user messages. Consider the router management code [in routeractor_group.go](https://github.com/AsynkronIT/protoactor-go/blob/dev/router/routeractor_group.go#L22-L29). It clearly expects the return value of `GetRoutees` to be a clone of the underlying `PIDSet`, otherwise it wouldn't have to call `SetRoutees` after the modification.

**Comment by rogeralsing** [comment]
ah doh. damn. I get it. the set is mutated in the actor and concurrently there is a user message routed trying to access the state of it.

**Comment by PotterDai** [comment]
@rogeralsing  Yep, that's the case.
@bryanpkc  I'm missing your point.
If `BroadcastMessage` is rare, then performance is not an issue here. Why don't we put it to mailbox?

**Comment by rogeralsing** [comment]
ok, I'm with @bryanpkc here, I'd rather see slow Add/Remove due to copy than slow user message routing.

**Comment by rogeralsing** [comment]
the main usecase for routers is to make something faster at delivering message to multiple other actors. so the default path of delivering fast must be respected.

in the case you need heavy mutation of the target actors, I would personally just use an actor for that. which manages both delivering and mutation of its own set of target actors.

**Comment by bryanpkc** [comment]
@PotterDai Router management messages, including `BroadcastMessage`, are indeed queued in the mailbox of the router management actor. I am saying that user messages (intended for routees) should not be queued up.

**Comment by PotterDai** [comment]
@bryanpkc @rogeralsing 
But we still need a RWMutex [here](https://github.com/AsynkronIT/protoactor-go/blob/dev/router/broadcast_router.go#L14) right?
Since SetRoutee and RouteMessage will still occur concurrently.

**Comment by rogeralsing** [comment]
IMO. copy on write for add/remove. 
that way, routing messages is exactly as fast as before. while add and remove will be affected by copy pidset performance. 

adding an RWMutex on the state would affect the routing performance.
most routers will never mutate their target set.

The scenario described in the original post is a special rare case.
Which very likely works just as good if all of the behavior was moved into an management actor.

**Comment by bryanpkc** [comment]
@PotterDai It would be sufficient for `SetRoutee` to call `atomic.StorePointer` on `&routees`. A `RWMutex` would be too heavyweight.

**Comment by PotterDai** [comment]
@bryanpkc Got it. Agree!

**Comment by NathanBaulch** [comment]
Should a similar fix be applied to `randomRouterState` and `roundRobinState`?

**Comment by rogeralsing** [comment]
fixed by #230
